utils/wasm-gen/src/generator/syscalls/invocator.rs

```diff
@@ -173,43 +178,156 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
             self.unstructured.len()
         );
 
-        for (invocable, (amount, call_indexes_handle)) in self.sys_call_imports.clone() {
+        self.insert_sys_calls()?;
+
+        log::trace!(
+            "Random data after inserting all sys-calls invocations - {}",
+            self.unstructured.len()
+        );
+
+        self.resolves_calls_indexes();
+
+        Ok(DisabledSysCallsInvocator {
+            module: self.module,
+            call_indexes: self.call_indexes,
+        })
+    }
+
+    fn insert_sys_calls(&mut self) -> Result<()> {
+        log::trace!(
+            "Random data before inserting sys-calls invoke instructions - {}",
+            self.unstructured.len()
+        );
+
+        let code_funcs = self.module.count_code_funcs();
+        let insert_into_funcs: Vec<_> = (0..code_funcs)
+            .filter(|idx| !self.call_indexes.is_custom_func(*idx))
+            .collect();
+
+        let syscalls_to_insert =
+            self.sys_call_imports
+                .clone()
+                .into_iter()
+                .flat_map(|(syscall, (amount, _))| {
+                    iter::repeat(syscall)
+                        .take(amount as usize)
+                        .collect::<Vec<_>>()
+                });
+
+        let insertion_mapping =
+            self.build_syscalls_insertion_mapping(syscalls_to_insert, &insert_into_funcs)?;
+
+        for (insert_into_fn, syscalls) in insertion_mapping {
+            self.insert_sys_calls_into_fn(insert_into_fn, syscalls)?;
+        }
+
+        log::trace!(
+            "Random data after inserting sys-calls invoke instructions - {}",
+            self.unstructured.len()
+        );
+
+        Ok(())
+    }
+
+    /// Distributes provided syscalls among provided function ids.
+    ///
+    /// Returns mapping `func_id` <-> `syscalls which should be inserted into func_id`.
+    fn build_syscalls_insertion_mapping<I>(
+        &mut self,
+        syscalls: I,
+        insert_into_funcs: &[usize],
+    ) -> Result<BTreeMap<usize, Vec<InvocableSysCall>>>
+    where
+        I: Iterator<Item = InvocableSysCall>,
+    {
+        let mut insertion_mapping: BTreeMap<usize, Vec<InvocableSysCall>> = BTreeMap::new();
```

Comment:
type can be auto inferred
```suggestion
        let mut insertion_mapping = BTreeMap::new();
```